app/waf/src/controllers/declaration.controller.ts

```diff
@@ -423,4 +432,85 @@ export class DeclarationController extends BaseController {
       },
     );
   }
+
+  async assertVirtualAddressAvailability(
+    adc: Adc,
+    declaration: AS3Declaration,
+  ): Promise<void> {
+    try {
+      let netId = (() => {
+        for (let netName of Object.keys(adc.networks)) {
+          let net = adc.networks[netName];
+          if (net.type === 'ext') return net.networkId;
+        }
+        throw new Error(
+          `not found the external network from adc: ${adc.networks}`,
+        );
+      })();
+
+      let netHelper = await this.wafapp.get(WafBindingKeys.KeyNetworkDriver);
+
+      let userToken = await this.reqCxt.get(
+        WafBindingKeys.Request.KeyUserToken,
+      );
+      let ports = await netHelper.getPortsInfo(userToken, netId);
+
+      let addresses = findByKey(declaration, 'virtualAddresses');
+      if (addresses.length === 0)
+        throw new Error('not found virtual servers, absurd declaration.');
+      for (let addrset of addresses) {
+        for (let addr of <string[]>addrset) {
+          // check addr on netId is available.
+
+          for (let port of ports) {
```

Comment:
also seems to me there is a loophole here? is it possible that the address is already occupied by another tenant because this is only checking against this tenant's ports? (scenario:  network shared by many tenants)

Reply:
good point.